File: src/io/ziheng/hashtable/leetcode/RansomNote.java
package io.ziheng.hashtable.leetcode;

import java.util.Map;
import java.util.HashMap;

/**
 * LeetCode 383. Ransom Note
 * https://leetcode.com/problems/ransom-note/
 */
public class RansomNote {
    public boolean canConstruct(String ransomNote, String magazine) {
        if (ransomNote == null || magazine == null) {
            return false;
        }
        return canConstructHashing(ransomNote, magazine);
    }
    public boolean canConstructHashing(String ransomNote, String magazine) {
        Map<Character, Integer> aMap = new HashMap<>();
        for (char c : magazine.toCharArray()) {
            aMap.put(c, 1 + aMap.getOrDefault(c, 0));
        }
        for (char c : ransomNote.toCharArray()) {
            if (!aMap.containsKey(c) || aMap.get(c) <= 0) {
                return false;
            }
            aMap.put(c, aMap.get(c) - 1);
        }
        return true;
    }
}
/* EOF */

File: src/cc/hashtable/RansomNote.cc
#include <map>
#include <string>

using namespace std;

/**
 * LeetCode 383. Ransom Note
 * https://leetcode.com/problems/ransom-note/
 */
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        map<char, int> magazineMap = buildMap(magazine);
        for (int i = 0; i < ransomNote.size(); i++) {
            char k = ransomNote[i];
            if (magazineMap.find(k) != magazineMap.end()
                && magazineMap[k] != 0) {
                magazineMap[k] -= 1;
            } else {
                return false;
            }
        }
        return true;
    }
private:
    map<char, int> buildMap(const string &str) {
        map<char, int> rMap = map<char, int>();
        for (int i = 0; i < str.size(); i++) {
            char k = str[i];
            if (rMap.find(k) == rMap.end()) {
                rMap[k] = 1;
            } else {
                rMap[k] += 1;
            }
        }
        return rMap;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}
/* EOF */